src/DotLiquid/Tokenizer.cs

```diff
@@ -124,32 +194,100 @@ private static void ReadToEnd(StringBuilder sb, DotLiquid.Util.CharEnumerator ma
                     switch (nextChar)
                     {
                         case '\'':
-                            isInQuotes = !isInQuotes;
-                            searchChars = isInQuotes ? SearchSingleQuoteEnd : initialSearchChars;
-                            break;
                         case '"':
-                            isInQuotes = !isInQuotes;
-                            searchChars = isInQuotes ? SearchDoubleQuoteEnd : initialSearchChars;
+                            ReadToChar(sb, markupEnumerator, nextChar);
                             break;
                         case '}':
                         case '%':
                             if (markupEnumerator.Remaining > 0 && markupEnumerator.Next == '}')
                             {
+                                var previousCharIsWhitespaceControl = syntaxCompatibilityLevel >= SyntaxCompatibility.DotLiquid22 && markupEnumerator.Previous == '-';
                                 markupEnumerator.AppendNext(sb);
-                                return;
+                                if (previousCharIsWhitespaceControl)
+                                {
+                                    // Remove hyphen from token
+                                    sb.Remove(sb.Length - 3, 1);
+
+                                    // Trim trailing whitespace by skipping ahead beyond the tag end
+                                    while (markupEnumerator.Remaining > 0)
+                                    {
+                                        if (((uint)markupEnumerator.Next - '\t') <= 5 || markupEnumerator.Next == ' ')
+                                            markupEnumerator.MoveNext();
+                                        else
+                                            break;
+                                    }
+                                }
+                                return true;
                             }
                             break;
                     }
                 }
             };
 
-            if (markupEnumerator.Remaining == 0) //Somehow we reached the end without finding the end character(s)
+            //Somehow we reached the end without finding the end character(s)
+            return false;
+        }
+
+        /// <summary>
+        /// Reads a token until, and inclusive of, the end character and advances the enumerator position
+        /// </summary>
+        /// <param name="sb">The StringBuilder to write to</param>
+        /// <param name="markupEnumerator">The string enumerator</param>
+        /// <param name="endChar">The character that indicates end of token</param>
+        /// <returns>True if reaches endChar, otherwise false</returns>
```

Comment:
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/documentation-comments#paramref
and
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/xmldoc/recommended-tags#see
```suggestion
        /// <returns><see langword="true"/> if reaches <paramref name="endChar"/>, otherwise <see langword="false"/></returns>
```